File: test/tutorial/test_tutorial.cpp
// Copyright Contributors to the libdnf project.
// SPDX-License-Identifier: GPL-2.0-or-later
//
// This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
//
// Libdnf is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// Libdnf is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with libdnf.  If not, see <https://www.gnu.org/licenses/>.


#include "test_tutorial.hpp"

#include <libdnf5/base/base.hpp>
#include <libdnf5/base/goal.hpp>
#include <libdnf5/repo/package_downloader.hpp>
#include <libdnf5/rpm/package_query.hpp>


CPPUNIT_TEST_SUITE_REGISTRATION(TutorialTest);


void TutorialTest::setUp() {
    temp = std::make_unique<libdnf5::utils::fs::TempDir>("libdnf_unittest");
    installroot = temp->get_path() / "installroot";
    cachedir = installroot / "var/cache/dnf/";
}


void TutorialTest::tearDown() {}


void TutorialTest::test_create_base() {
#include "session/create_base.cpp"
}


void TutorialTest::test_load_repo() {
#include "session/create_base.cpp"

    base.get_config().get_cachedir_option().set(cachedir);

#include "repo/load_repo.cpp"
}


void TutorialTest::test_load_system_repos() {
#include "session/create_base.cpp"

    base.get_config().get_cachedir_option().set(cachedir);

#include "repo/load_system_repos.cpp"
}


void TutorialTest::test_query() {
#include "session/create_base.cpp"

    base.get_config().get_cachedir_option().set(cachedir);

#include "query/query.cpp"
#include "repo/load_repo.cpp"
}


void TutorialTest::test_transaction() {
#include "session/create_base.cpp"

    base.get_config().get_cachedir_option().set(cachedir);

#include "repo/load_repo.cpp"
#include "transaction/transaction.cpp"
}


void TutorialTest::test_force_arch() {
#include "session/force_arch.cpp"
}

File: test/python3/libdnf5/tutorial/test_tutorial.py
# Copyright Contributors to the libdnf project.
#
# This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
#
# Libdnf is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# Libdnf is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with libdnf.  If not, see <https://www.gnu.org/licenses/>.

import os
import pathlib
import tempfile

import base_test_case

PROJECT_BINARY_DIR = os.environ["PROJECT_BINARY_DIR"]
PROJECT_SOURCE_DIR = os.environ["PROJECT_SOURCE_DIR"]


class TestTutorial(base_test_case.BaseTestCase):
    def setUp(self):
        super().setUp()
        self.installroot = self.base.get_config().installroot
        self.cachedir = self.base.get_config().cachedir
        self.baseurl = pathlib.Path(os.path.join(
            PROJECT_BINARY_DIR, "test/data/repos-rpm/rpm-repo1/")).as_uri()

    def test_create_base(self):
        file = ""
        with open("tutorial/session/create_base.py", "r") as f:
            file += f.read()

        exec(file, {'installroot': self.installroot, 'cachedir': self.cachedir})

    def test_load_repo(self):
        file = ""
        with open("tutorial/session/create_base.py", "r") as f:
            file += f.read()

        with open("tutorial/repo/load_repo.py", "r") as f:
            file += f.read()

        exec(file, {'installroot': self.installroot,
                    'cachedir': self.cachedir, 'baseurl': self.baseurl})

    def test_load_system_repo(self):
        # TODO(nsella) This example does not 'compile' yet
        # file = ""
        # with open("tutorial/session/create_base.py", "r") as f:
        #     file += f.read()

        # with open("tutorial/repo/load_system_repos.py", "r") as f:
        #     file += f.read()

        # exec(file, { 'installroot': self.installroot, 'baseurl': self.baseurl })
        pass

    def test_query(self):
        file = ""
        with open("tutorial/session/create_base.py", "r") as f:
            file += f.read()

        with open("tutorial/repo/load_repo.py", "r") as f:
            file += f.read()

        with open("tutorial/query/query.py", "r") as f:
            file += f.read()

        exec(file, {'installroot': self.installroot,
                    'cachedir': self.cachedir, 'baseurl': self.baseurl})

    def test_transaction(self):
        file = ""
        with open("tutorial/session/create_base.py", "r") as f:
            file += f.read()

        with open("tutorial/repo/load_repo.py", "r") as f:
            file += f.read()

        with open("tutorial/transaction/transaction.py", "r") as f:
            file += f.read()

        tmp_path = os.getcwd()
        exec(file, {'installroot': self.installroot,
                    'cachedir': self.cachedir, 'baseurl': self.baseurl})
        os.chdir(tmp_path)

    def test_force_arch(self):
        file = ""
        with open("tutorial/session/force_arch.py", "r") as f:
            file += f.read()

        exec(file, {'installroot': self.installroot, 'cachedir': self.cachedir})
